examples_tests/17.SimpleBulletIntegration/Camera.hpp

```diff
@@ -0,0 +1,377 @@
+
+	inline float getFarValue() const { return zFar; }
+
+	inline float getAspectRatio() const { return aspectRatio; }
+
+	inline float getFOV() const { return fovy; }
+
+	inline void setLeftHanded(bool _leftHanded = true)
+	{
+		leftHanded = _leftHanded;
+		recomputeViewMatrix();
+		recomputeProjectionMatrix();
+	}
+
+	inline void setNearValue(float zn)
+	{
+		zNear = zn;
+		recomputeProjectionMatrix();
+	}
+
+	inline void setFarValue(float zf)
+	{
+		zFar = zf;
+		recomputeProjectionMatrix();
+	}
+
+	inline void setAspectRatio(float aspect)
+	{
+		aspectRatio = aspect;
+		recomputeProjectionMatrix();
+	}
+
+	inline void setFOV(float fovy)
+	{
+		fovy = fovy;
+		recomputeProjectionMatrix();
+	}
+
+public:
+
+	void mouseProcess(const IMouseEventChannel::range_t& events)
+	{
+		for (auto eventIt=events.begin(); eventIt!=events.end(); eventIt++)
+		{
+			auto ev = *eventIt;
+
+			if(ev.type == ui::SMouseEvent::EET_CLICK && ev.clickEvent.mouseButton == ui::EMB_LEFT_BUTTON) {
+				if(ev.clickEvent.action == ui::SMouseEvent::SClickEvent::EA_PRESSED) {
+					mouseDown = true;
+				} else if (ev.clickEvent.action == ui::SMouseEvent::SClickEvent::EA_RELEASED) {
+					mouseDown = false;
+				}
+			}
+
+			if(ev.type == ui::SMouseEvent::EET_MOVEMENT && mouseDown) {
+				float rotateSpeed = 0.2f;
+				core::vectorSIMDf pos; pos.set(getPosition());
+				core::vectorSIMDf target = getTarget() - pos;
+				core::vector3df relativeRotation = target.getAsVector3df().getHorizontalAngle();
+				relativeRotation.X -= ev.movementEvent.movementY * rotateSpeed * -1.0f;
+				float tmpYRot = ev.movementEvent.movementX * rotateSpeed * -1.0f;
+				if (leftHanded)
+					relativeRotation.Y -= tmpYRot;
+				else
+					relativeRotation.Y += tmpYRot;
+
+				constexpr float MaxVerticalAngle = 88.0f;
+				if (relativeRotation.X > MaxVerticalAngle*2 &&
+					relativeRotation.X < 360.0f-MaxVerticalAngle)
+				{
+					relativeRotation.X = 360.0f-MaxVerticalAngle;
+				}
+				else
+				if (relativeRotation.X > MaxVerticalAngle &&
+					relativeRotation.X < 360.0f-MaxVerticalAngle)
+				{
+					relativeRotation.X = MaxVerticalAngle;
+				}
+
+				target.set(0,0, core::max(1.f, core::length(pos)[0]), 1.f);
+				core::vectorSIMDf movedir = target;
+
+				core::matrix3x4SIMD mat;
+				{
+					core::matrix4x3 tmp;
+					tmp.setRotationDegrees(core::vector3df(relativeRotation.X, relativeRotation.Y, 0));
+					mat.set(tmp);
+				}
+				mat.transformVect(target);
+				
+				// write right target
+				target += pos;
+				setTarget(target.getAsVector3df());
+			}
+		}
+	}
+
+	void keyboardProcess(const IKeyboardEventChannel::range_t& events)
+	{
+		double perActionDt[Keys::EKA_COUNT] = {};
+
+		for(uint32_t k = 0; k < Keys::EKA_COUNT; ++k) {
+			if(keysDown[k] == true) {
+				auto timeDiff = std::chrono::duration_cast<std::chrono::milliseconds>(nextPresentationTimeStamp - lastVirtualUpTimeStamp).count();
+				assert(timeDiff >= 0);
+				perActionDt[k] += timeDiff;
+			}
+		}
```

Comment:
this a very clever solution, took me a while to understand, would be useful to have comments about the reasoning